src/libraries/TwabLib.sol

```diff
@@ -8,208 +8,119 @@ import "./RingBufferLib.sol";
-  /// @return accountDetails The new AccountDetails
-  /// @return twab The user's latest TWAB
-  /// @return isNew Whether the TWAB is new
-  function increaseBalance(
+  function increaseBalances(
     Account storage _account,
     uint112 _amount,
-    uint32 _currentTime
-  )
-    internal
-    returns (
-      AccountDetails memory accountDetails,
-      ObservationLib.Observation memory twab,
-      bool isNew
-    )
-  {
-    AccountDetails memory _accountDetails = _account.details;
-    (accountDetails, twab, isNew) = _nextTwab(_account.twabs, _accountDetails, _currentTime);
-    accountDetails.balance = _accountDetails.balance + _amount;
+    uint112 _delegateAmount
+  ) internal returns (ObservationLib.Observation memory twab, bool isNewTwab) {
+    if (_delegateAmount != uint112(0)) {
+      (twab, isNewTwab) = _nextTwab(_account);
+    }
+
+    _account.balance += _amount;
+    _account.delegateBalance += _delegateAmount;
   }
 
   /**
-   * @notice Calculates the next TWAB checkpoint for an account with a decreasing balance.
+   * @notice Calculates the next TWAB checkpoint for an account with a decreasing delegateBalance.
    * @dev    With Account struct and amount decreasing calculates the next TWAB observable checkpoint.
-   * @param _account        Account whose balance will be decreased
-   * @param _amount         Amount to decrease the balance by
-   * @param _revertMessage  Revert message for insufficient balance
-   * @return accountDetails Updated Account.details struct
+   * @param _account        Account whose delegateBalance will be decreased
+   * @param _amount         Amount to decrease the delegateBalance by
+   * @param _delegateAmount   The amount to increase the delegateBalance by
+   * @param _revertMessage  Revert message for insufficient delegateBalance
    * @return twab           TWAB observation (with decreasing average)
-   * @return isNew          Whether TWAB is new or calling twice in the same block
+   * @return isNewTwab          Whether TWAB is new or calling twice in the same block
    */
-  function decreaseBalance(
+  function decreaseBalances(
     Account storage _account,
     uint112 _amount,
-    string memory _revertMessage,
-    uint32 _currentTime
-  )
-    internal
-    returns (
-      AccountDetails memory accountDetails,
-      ObservationLib.Observation memory twab,
-      bool isNew
-    )
-  {
-    AccountDetails memory _accountDetails = _account.details;
-
-    require(_accountDetails.balance >= _amount, _revertMessage);
-
-    (accountDetails, twab, isNew) = _nextTwab(_account.twabs, _accountDetails, _currentTime);
+    uint112 _delegateAmount,
+    string memory _revertMessage
+  ) internal returns (ObservationLib.Observation memory twab, bool isNewTwab) {
+    require(_account.balance >= _amount, _revertMessage);
+    require(_account.delegateBalance >= _delegateAmount, _revertMessage);
+
+    if (_delegateAmount != uint112(0)) {
+      (twab, isNewTwab) = _nextTwab(_account);
```

Comment:
Now the `nextTwabIndex` and `cardinality` are updated separately from the balance and delegated balance.  This looks like it will use 2 SSTORES now